### ISSUE TITLE
Distributed PresCat / ManyCats

### ISSUE THREAD

**Issue by julianmorley** [sul-dlss/preservation_catalog#1190]
Future storage for SDR preservation is now likely to be in the form of stand-alone, high capacity servers rather than via centralized storage over NFS. What changes must be made to PresCat to support multiple instances of the PresCat stack, with each stack running on a stand-alone server, responsible only for content on the local hard drives, and unaware of any other Prescat instances?

**Comment by jmartin-sul** [comment]
if there were no need for a central catalog with everything across the subcatalogs... i think prescat would already support this?

unless the new storage approach didn't put the moabs on a posix file system.  there would need to be some significant reworking, i'd think, if the posix assumption no longer held.  which would also require changes to (or replacement of) the moab-versioning gem.

e.g. with the current code, you could imagine an individual prescat instance for each service disk.  though at present that would also mean a postgres instance and a redis instance for each such prescat instance.

my guess is that this is the sort of thing that'd need a more dedicated work cycle, and that it'd be hard to fit it into e.g. a 5ish week maintenance work cycle.

**Comment by jmartin-sul** [comment]
> if there were no need for a central catalog with everything across the subcatalogs

but fwiw, i think a central catalog of everything would be a good thing.  and so i think this would take some work to do it "right".

**Comment by vivnwong** [comment]
Any major enhancement should be in the preservation focused work cycle.

**Comment by julianmorley** [comment]
I think there might be some work around the ingest WF for versioning objects that are not on the most recent services disk. Or maybe we abandon the WF hook for new deposits and just rely on internal Prescat disk audits to detect new content in a timely manner.

**Comment by julianmorley** [comment]
I agree with @jmartin-sul that current presCat would work as-is; we'd just lose a central catalog of all content across all presCat nodes, and we'd potentially lose immediate preservation for new versions of older objects (Moabs that are not on the most recent services disk).

The intent would be to run the full stack on each SDR brick; no big deal in terms of resources, as these devices are each equivalent to an VMWare host machine and so will have 20+ cores and 192GB+ RAM available exclusively for PresCat; more than enough for a Postgres db, Redis, plus enough workers to handle audit and transfer duties.

**Comment by tallenaz** [comment]
Per @vivnwong's request for progress on this front:
- We've provisioned a standalone, high capacity server with Ubuntu and ZFS installed.
- We have puppet manifests to set it up with Rails, Postgres10, and Redis, and a branch of preservation_catalog that deploys all of those components of the application stack to the server. 
- We've tested several rounds of audit checks, in particular, MoabToCatalog and CatalogToMoab, verifying that ZFS is a viable storage solution.
- We're in process with setting up and testing workers for the various cloud endpoints. 
- Finally, we've installed mpifileutils, to allow faster file transfers from the existing storage roots to ZFS, which will reduce migration time.

**Comment by jmartin-sul** [comment]
based on a meeting earlier this week, the plan is for @ndushay and i to take a shot at describing and estimating LoE for the remaining development work to do this (via an [ADR](https://github.com/sul-dlss-labs/infrastructure-adrs/)).

**Comment by jmartin-sul** [comment]
current state of this:
* the immediate priority is supporting multiple moabs per druid.  while not strictly necessary to support the distributed version of pres cat described in the issue summary, we've come to the conclusion that this change would allow for easier and less disruptive migration of content between storage locations.  this is both desirable in and of itself, and helpful for the practical running of the distributed pres cat future this ticket describes.
  * we had a very useful meeting about this, with notes here, which i'll try to summarize more pithily in a follow up comment:  https://docs.google.com/document/d/1D84sbcLmldD7UY3IsHe9s5tp_Q6AJMXQfMPWrEsROHM/edit#heading=h.bk27kxkbyztz
* i've altered the title to more accurately describe what we're planning to do, and in what order, while retaining a cute cat pun to the best of my ability.  i believe "distributed pres cat" is more accurate than "ManyCats", as we're planning on continuing to run a single centralized pres cat database, albeit with an eye towards having each pres cat instance deal only with moabs on its locally mounted storage, even if we only get to the "1+ moab per druid" part of things in this work cycle.  that would still get us to the goal of "PresCat [supporting] multiple instances of the PresCat stack, with each [instance] running on a stand-alone server, responsible only for content on the local hard drives, and unaware of any other Prescat instances".
  * fwiw, i suspect that it will be easier and slightly less work to change the code so that a given pres cat instance only cares about a subset of all the storage.

**Comment by ndushay** [comment]
I'm closing this multi-moab EPIC.  If that's undesired, please re-open.
